File: dataFormatting/songnode.cpp


#ifndef SONGNODE_CPP
#define SONGNODE_CPP


#include <regex>
#include <map>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>
#include <utility>
#include <vector>
#include <unordered_set>

using namespace std;


struct SongNode
{



string _name;
string _artist;
string _album;

map<string, int> _shared;



//
string getName()
{

	return this->_name;

}


//
void printName()
{

	cout << this->_name << '\n';

}


//
string getArtist()
{

	return this->_artist;

}


//
void printArtist()
{

	cout << this->_artist << '\n';

}


//
string getAlbum()
{

	return this->_album;

}


//
void printAlbum()
{

	cout << this->_album << '\n';

}


//
map<string, int> getSongs()
{

	return this->_shared;

}


//
void printSongs()
{

	cout << "\nSong : Count\n\n";


	for (auto iter = this->_shared.begin(); iter != this->_shared.end(); iter++)
	{

		cout << get<0>(*iter) << " : " << get<1>(*iter) << endl;

	}


}


//
bool findSong(string song)
{

	auto found = this->_shared.find(song);


	if (found == this->_shared.end())
	{

		return false;

	}

	else
	{
		return true;
	}

}


//
int getCount(string songname)
{

	auto found = this->_shared.find(songname);

	if (found != this->_shared.end())
	{
		return get<1>(*found);
	}
	else
	{
		return 0;
	}

}


//
static bool valueCompare(pair<string, int>& lhs, pair<string, int>& rhs)
{

	return lhs.second < rhs.second;

}


//
unordered_set<string> getHighest(int num = 1)
{


	vector<pair<string, int>> list;

	unordered_set<string> output;


	for (auto iter = this->_shared.begin(); iter != this->_shared.end(); iter++)
	{

		list.push_back(pair(*iter));

	}

	sort(list.begin(), list.end(), valueCompare);

	
	if ( list.size() < num)
	{
		num = list.size();
	}


	for (int index = list.size() - num; index < list.size(); index++)
	{

		output.insert(get<0>(list[index]));

	}


	return output;

}


void addShared(string shared)
{

	if (this->findSong(shared))
	{
		this->_shared[shared]++;
	}
	else
	{
		this->_shared[shared] = 1;
	}

}


static string parseName(string name)
{


	regex next("\\\\");
	
	name = regex_replace(name, next, "!BSL");


	next = "/";

	name = regex_replace(name, next, "!FSL");


	next = ",";

	name = regex_replace(name, next, "!CMA");


	next = "\"";

	name = regex_replace(name, next, "`");


	next = "'";

	name = regex_replace(name, next, "`");


	next = "\\?";

	name = regex_replace(name, next, "!QST");


	next = "\\*";

	name = regex_replace(name, next, "!AST");


	next = "\\|";

	name = regex_replace(name, next, "!BAR");


	next = ":";

	name = regex_replace(name, next, "!SMI");


	next = "<";

	name = regex_replace(name, next, "!LNG");


	next = ">";

	name = regex_replace(name, next, "!RNG");


	next = "\\.";

	name = regex_replace(name, next, "!PRD");


	// Trailing whitespace
	next = "\\s+$";

	name = regex_replace(name, next, "");


	if (251 > name.size())
	{

		name = rmvPunct(name);

	}


	return name;


}


static string parseReverse(string name)
{

	regex next("!BSL");

	name = regex_replace(name, next, "\\");

	next = "!FSL";

	name = regex_replace(name, next, "/");

	next = "!CMA";

	name = regex_replace(name, next, ",");

	next = "`";

	name = regex_replace(name, next, "'");

	next = "!QST";

	name = regex_replace(name, next, "?");

	next = "!AST";

	name = regex_replace(name, next, "*");

	next = "!BAR";

	name = regex_replace(name, next, "|");

	next = "!SMI";

	name = regex_replace(name, next, ":");

	next = "!LNG";

	name = regex_replace(name, next, "<");

	next = "!RNG";

	name = regex_replace(name, next, ">");

	next = "!PRD";

	name = regex_replace(name, next, ".");


	return name;


}


static string rmvCommas(string input)
{

	regex next("\\\\");

	input = regex_replace(input, next, "/");

	next = ",";

	input = regex_replace(input, next, ";");

	next = "'";

	input = regex_replace(input, next, "`");

	next = "\"";

	input = regex_replace(input, next, "`");


	return input;

}


void readCsv(string songname)
{

	this->_name = songname;

	string filepath = "tracks/" + songname + ".csv";


	ifstream file;

	file.open(filepath);


	string line_string;

	getline(file, line_string, '\n');


	istringstream line_stream(line_string);

	string next;

	getline(line_stream, next, ',');


	this->_artist = next;


	getline(line_stream, next);

	this->_album = next;


	getline(file, line_string, '\n');

	string count;


	while (!line_string.empty())
	{
	
		line_stream = istringstream(line_string);

		getline(line_stream, next, ',');

		getline(line_stream, count);

		if (!count.empty())
		{
			this->_shared[next] = stoi(count);
		}


		getline(file, line_string, '\n');
		

	}

	file.close();


}


void toCsv(bool testing = false)
{

	string filepath = this->_name + ".csv";


	if (!testing)
	{

		filepath = "tracks/" + filepath;

	}


	ofstream file(filepath, ios::out | ios::trunc);

	file << this->_artist << ',' << this->_album << endl;

	
	for (auto iter = this->_shared.begin(); iter != this->_shared.end(); iter++)
	{

		file << get<0>(*iter) << ',' << get<1>(*iter) << endl;

	}


	file.close();


}


static string rmvPunct(string name)
{

	regex next("!BSL");

	name = regex_replace(name, next, "");

	next = "!FSL";

	name = regex_replace(name, next, "");

	next = "!CMA";

	name = regex_replace(name, next, "");

	next = "`";

	name = regex_replace(name, next, "");

	next = "!QST";

	name = regex_replace(name, next, "");

	next = "!AST";

	name = regex_replace(name, next, "");

	next = "!BAR";

	name = regex_replace(name, next, "");

	next = "!SMI";

	name = regex_replace(name, next, "");

	next = "!LNG";

	name = regex_replace(name, next, "");

	next = "!RNG";

	name = regex_replace(name, next, "");

	next = "!PRD";

	name = regex_replace(name, next, "");


	return name;


}


static SongNode load(string songname)
{

	SongNode rhs;

	rhs.readCsv(songname);

	return rhs;

}




};



#endif // SONGNODE_CPP

File: dataFormatting/songnode.py
import boto3
import pandas as pd
from io import StringIO
import re
import os
import json

class SongNode :


    def __init__(self, name = "", artist = "", album = ""):

        self._name = name
        self._artist = artist
        self._album = album
        self._shared = dict()


    def get_name(self):

        return self._name


    def print_name(self) :

        print(self.get_name())


    def get_artist(self) :

        return self._artist


    def print_artist(self) :

        print(self.get_artist())


    def get_album(self) :
    
        return self._album


    def print_album(self) :

        print(self.get_album())


    def get_songs(self):

        return self._shared


    def print_songs(self) :

        print(self.get_songs())


    def find_song(self, song) :

        return song in self._shared.keys()


    def get_count(self, song) :


        if self.find_song(song) :
    
            return self._shared[song]

        else :

            return 0

    

    def get_highest(self, num=1) :

        return [x[0] for x in reversed(sorted(self.get_songs().items(), key = lambda y : y[1]))][0 : num]
        

    def add_shared(self, shared):
    

        if self.find_song(shared) :

            self._shared[shared] = int(self._shared[shared]) + 1

        else :

            self._shared[shared] = 1


    @staticmethod
    def parse_name(name) :

        
        # Parsing out illegal filename characters


        replacements = {
                        '\\' : '!BSL',
                        '/' : '!FSL',
                        ',' : '!CMA',
                        "'" : '`',
                        '"' : '`',
                        '?' : '!QST',
                        '*' : '!AST',
                        '|' : '!BAR',
                        ':' : '!SMI',
                        '<' : '!LNG',
                        '>' : '!RNG',
                        '.' : '!PRD'
                        }

        regex = re.compile("%s" % "|".join( map( re.escape, replacements.keys() ) ) )
                        
        output = regex.sub( lambda found : str( replacements[found.group()] ), str( name ) )
        output = output.strip()


        if len(output) > 251 :

            output = SongNode.rmv_punct(output)


        return output

    @staticmethod
    def parse_reverse(name) :

        replacements = {
                        '!BSL' : '\\',
                        '!FSL' : '/',
                        '!CMA' : ',',
                        '`' : "'",
                        '!QST' : '?',
                        '!AST' : '*',
                        '!BAR' : '|',
                        '!SMI' : ':',
                        '!LNG' : '<',
                        '!RNG' : '>',
                        '!PRD' : '.'
                        }

        regex = re.compile("%s" % "|".join( replacements.keys() ) )
                        
        return  regex.sub( lambda found : replacements[found.group()], name )
       

    
    @staticmethod
    def rmv_commas(string) :


        replacements = {
                        '\\' : '/',
                        ',' : ';',
                        '"' : '`',
                        "'" : '`'
                        }

        regex = re.compile("%s" % "|".join( map(re.escape, replacements) ) )
        return regex.sub( lambda found : replacements[found.group()], string )



    def read_csv(self, songname, testing = False) :

        print("Current working directory:", os.getcwd())
        
        with open('../config.json', 'r') as config_file:
            config = json.load(config_file)

        session = boto3.Session(
            aws_access_key_id=config['aws_access_key_id'],
            aws_secret_access_key=config['aws_secret_access_key'],
            region_name=config['region_name']
        )

        s3 = session.client('s3')
        client = boto3.client('s3')
        
        bucket_name = 'tracksdataset'
        paginator = s3.get_paginator('list_objects_v2')
        pages = paginator.paginate(Bucket=bucket_name)
        objKeys = []
        
        found = False
        
        for page in pages:
            for obj in page.get('Contents', []):
                if(obj["Key"].replace('.csv', '') == songname):
                    objKeys.append(obj["Key"])
                    found = True
                    break
            if found:
                break
        
        csv_obj = client.get_object(Bucket=bucket_name, Key=objKeys[0])
        
        body = csv_obj['Body']
        csv_string = body.read().decode('utf-8')

        df = pd.DataFrame(StringIO(csv_string))

        columns = list(df)
            
        for i, j in df.iterrows():
            if(i == 0):
                continue
            s = ""
            for k in columns:
                s += df[k][i]
            p = s.split(",")
            self._shared[p[0]] = int(p[1].replace('\r\n', ''))
        


    def to_csv(self, filepath = None, testing = False) :

        name = self.get_name()

        #name = SongNode.parse_name(name)


        if (False == testing) and (None == filepath) :

            filepath = 'tracks/' + name + '.csv'

        elif (True == testing) and (None == filepath) :

            filepath = name + '.csv'

        
        
        file = open(filepath, mode='w', encoding='utf-8')

        file.write(self._artist + "," + self._album + '\n')

        
        for song, count in self._shared.items() :

            file.write(song + ',' + str( count ) + '\n')

        file.close()


    @staticmethod
    def rmv_punct(self, string) :

 
        replacements = {
                        '!BSL' : '',
                        '!FSL' : '',
                        '!CMA' : '',
                        '`' : "",
                        '!QST' : '',
                        '!AST' : '',
                        '!BAR' : '',
                        '!SMI' : '',
                        '!LNG' : '',
                        '!RNG' : '',
                        '!PRD' : ''
                        }

        regex = re.compile("%s" % "|".join( map(re.escape, replacements) ) )

        return regex.sub( lambda found : replacements[found.group()], string )
        


    @staticmethod
    def load(songname, testing = False) :

        lhs = SongNode()

        lhs.read_csv(songname, testing = testing)

        return lhs



